### PR TITLE
Update invoke logic to support running stateless actions without a claim

### LINKED ISSUE #821: Support running custom actions without installing the bundle
The spec allows for executing a stateless custom action when a bundle is not installed. 

> The stateless field indicates that the runtime will bypass credentials requirements (the user MAY or MAY NOT pass any credentials), and will not keep track of this action. If credentials are supplied, the runtime SHOULD pass the credentials to the invocation image (as in stateful actions). An invocation image MAY exit with an error if the necessary credentials are not supplied. However, the runtime MUST NOT fail to invoke the invocation image because of missing credentials.

> The primary intention of a stateless action is to allow invocation bundles to provide dry-run or detailed help functionalities. Stateless actions can be invoked on a non-existing installation. In this case, the name passed to the runtime need not exist, and MUST not result in the creation of a name record. Runtimes MAY ignore name if supplied.

However cnab-go is currently written to want a claim:

https://github.com/cnabio/cnab-go/blob/2fdc546bd3c7400a68b251c79831097fc1500c4d/action/run_custom.go#L30

Let's update Porter to make it work!

make it work

* Allow running custom actions when there isn't a claim.
* If the custom action is stateless, at the end of the action, a claim should not be persisted. Otherwise a claim should be persisted.
* upgrade still requires install to be run first.
* Since there is uncertainty about changing cnab-go's design (see related issues below), for this issue, let's workaround it's current design in Porter for now.

See related discussions in duffle-land
* https://github.com/cnabio/duffle/issues/830
* https://github.com/cnabio/duffle/pull/870

### PR DESCRIPTION
**Pull request by jeremyrickard**
Signed-off-by: Jeremy Rickard <jeremy.rickard@microsoft.com>

# What does this change

This PR updates Porter's internal invoke logic to support running stateless actions *without* having claim

```
jeremyrickard@MININT-1L9PE4A:stateless$ rm ~/.porter/claims/*
jeremyrickard@MININT-1L9PE4A:stateless$ porter invoke --action preflight whatsup
invoking custom action preflight on whatsup...
executing preflight action from HELLO (bundle instance: whatsup)
Check the world
The world is good
execution completed successfully!
jeremyrickard@MININT-1L9PE4A:stateless$ tree ~/.porter/claims/
/Users/jeremyrickard/.porter/claims/

0 directories, 0 files
jeremyrickard@MININT-1L9PE4A:stateless$ porter install
installing HELLO...
executing install action from HELLO (bundle instance: HELLO)
Install Hello World
Hello World
execution completed successfully!
jeremyrickard@MININT-1L9PE4A:stateless$ tree ~/.porter/claims/
/Users/jeremyrickard/.porter/claims/
└── HELLO.json

0 directories, 1 file
```

# What issue does it fix
Closes #821 

# Checklist
- [x] Unit Tests
- [ ] Documentation
  - [x] Documentation Not Impacted

swan song

### CODE REVIEW

File: pkg/cnab/provider/invoke.go [hunk 1]
```diff
@@ -4,25 +4,51 @@ import (
...
+	var err error
 	if args.BundlePath != "" {
-		c.Bundle, err = d.LoadBundle(args.BundlePath, args.Insecure)
+		bun, err = d.LoadBundle(args.BundlePath, args.Insecure)
 		if err != nil {
 			return err
 		}
 	}
 
+	c, tempClaim, err := d.getClaim(bun, action, args.Claim)
```
Comment (vdice) [inline pkg/cnab/provider/invoke.go]: super nit: maybe `isTemp` instead?  when scanning these lines I was thinking it was a temp claim object...

File: pkg/cnab/provider/invoke.go [hunk 2]
```diff
@@ -4,25 +4,51 @@ import (
...
 	cnabaction "github.com/deislabs/cnab-go/action"
+	"github.com/deislabs/cnab-go/bundle"
+	"github.com/deislabs/cnab-go/claim"
 	"github.com/hashicorp/go-multierror"
 	"github.com/pkg/errors"
 )
 
-func (d *Runtime) Invoke(action string, args ActionArguments) error {
-	c, err := d.instanceStorage.Read(args.Claim)
+func (d *Runtime) getClaim(bun *bundle.Bundle, actionName, claimName string) (*claim.Claim, bool, error) {
```
Comment (vdice) [inline pkg/cnab/provider/invoke.go]: Even though unexported, I wonder if we can write a function description here, especially to clarify the return vals?

File: pkg/cnab/provider/invoke.go [hunk 3]
```diff
@@ -4,25 +4,51 @@ import (
...
+	}
+	return &c, false, nil
+}
+
+func (d *Runtime) writeClaim(tempClaim bool, c *claim.Claim) error {
+	if !tempClaim {
+		return d.instanceStorage.Store(*c)
 	}
+	return nil
+}
```
Comment (vdice) [inline pkg/cnab/provider/invoke.go]: super nit: can we add a newline here?

File: pkg/cnab/provider/invoke_test.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
+package cnabprovider
+
+import (
+	"testing"
+
+	"get.porter.sh/porter/pkg/config"
+	"github.com/deislabs/cnab-go/bundle"
+	"github.com/deislabs/cnab-go/claim"
+	instancestorage "github.com/deislabs/porter/pkg/instance-storage"
```
Comment (vdice) [inline pkg/cnab/provider/invoke_test.go]: oh here's another import that needs updating

**Comment by jeremyrickard** [comment]
hey @vdice can you take one last look before i squash this?

**Pull request merged**